Creating a Fully Managed Beam Streaming System with Flink Runner on Kubernetes - Part 3 (#29860)
apache/beam: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md

---
title:  "Behind the Scenes: Crafting an Autoscaler for Apache Beam in a High-Volume Streaming Environment"

> date:   2023-12-21 09:00:00 -0400

Reviewer: Leaving a comment to remind us to update this before merging

categories:
  - blog
authors:
  - talat
---
<!--
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

### Introduction to the Design of Our Autoscaler for Apache Beam Jobs

> Welcome to the third and final part of our blog series on building a scalable, self-managed streaming infrastructure with Beam and Flink. In our previous post, we delved into the scale of our streaming platforms, highlighting our capacity to manage over 40,000 streaming jobs and process upwards of 10 million events per second. This impressive scale sets the stage for the challenge we address today: the intricate task of resource allocation in a dynamic streaming environment.

Reviewer: "In our previous post" - would benefit for an inline link to the post

> In this blog post [Talat Uyarer (Architect / Senior Principal Engineer)](https://www.linkedin.com/in/talatuyarer/), [Rishabh Kedia (Principal Engineer)](https://www.linkedin.com/in/rishabhkedia/) describe more details about our Autoscaler. Imagine a scenario where your streaming system is inundated with fluctuating workloads. Our case presents a unique challenge, as our customers, equipped with firewalls distributed globally, generate logs at various times of the day. This results in workloads that not only vary by time but also escalate over time due to changes in settings or the addition of new cybersecurity solutions from PANW. Furthermore, updates to our codebase necessitate rolling out changes across all streaming jobs, leading to a temporary surge in demand as the system processes unprocessed data.

Reviewer: ```suggestion
In this blog post [Talat Uyarer (Architect / Senior Principal Engineer)](https://www.linkedin.com/in/talatuyarer/) and [Rishabh Kedia (Principal Engineer)](https://www.linkedin.com/in/rishabhkedia/) describe more details about our Autoscaler. Imagine a scenario where your streaming system is inundated with fluctuating workloads. Our case presents a unique challenge, as our customers, equipped with firewalls distributed globally, generate logs at various times of the day. This results in workloads that not only vary by time but also escalate over time due to changes in settings or the addition of new cybersecurity solutions from PANW. Furthermore, updates to our codebase necessitate rolling out changes across all streaming jobs, leading to a temporary surge in demand as the system processes unprocessed data.
```

<img class="center-block"
src="/images/blog/apache-beam-flink-and-kubernetes-part3/resource-allocation.png"

> alt="Resource Allocation">

Reviewer: Nit: I think a more accurate/useful graphic might have some overlap between the orange and blue lines for the underprovisioned half. It is pretty rare to have a pipeline or service that is underprovisioned to the point of _never_ having resources, compared to not having enough resources to handle spikes

Reviewer: The graphic as drawn will never catch up

Author: I updated the image

Join us as we delve into the design and functionality of our autoscaler, a solution tailored to bring efficiency, adaptability, and a touch of intelligence to the world of streaming infrastructure.

## Identifying the Right Signals for Autoscaling

When we're overseeing a system like Apache Beam jobs on Flink, it's crucial to identify key signals that help us understand the relationship between our workload and resources. These signals are our guiding lights, showing us when we're lagging behind or wasting resources. By accurately identifying these signals, we can formulate effective scaling policies and implement changes in real-time. Imagine needing to expand from 100 to 200 TaskManagers — how do we smoothly make that transition? That's where these signals come into play.

Remember, we're aiming for a universal solution applicable to any workload and pipeline. While specific problems might benefit from unique signals, our focus here is on creating a one-size-fits-all approach.

> In Flink, tasks form the basic execution unit and consist of one or more operators, such as map, filter, or reduce. Flink optimizes performance by chaining these operators into single tasks when possible, minimizing overheads like thread context switching and network I/O. Your pipeline, when optimized, turns into a directed acyclic graph of stages, each processing elements based on your code. Don't confuse stages with physical machines — they're separate concepts. In Our job we mesuare backlog information by using Apache Beam's [`backlog_bytes` and `backlog_elements`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java#L32) metrics.

Reviewer: ```suggestion
In Flink, tasks form the basic execution unit and consist of one or more operators, such as map, filter, or reduce. Flink optimizes performance by chaining these operators into single tasks when possible, minimizing overheads like thread context switching and network I/O. Your pipeline, when optimized, turns into a directed acyclic graph of stages, each processing elements based on your code. Don't confuse stages with physical machines — they're separate concepts. In our job we measure backlog information by using Apache Beam's [`backlog_bytes` and `backlog_elements`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java#L32) metrics.
```

#### **Downscaling: When to Reduce Resources**

Downscaling is like knowing when some chefs can take a break after a rush hour. We consider this when:

- Our backlog is low — we've caught up with the orders.
- The backlog isn't growing — we're keeping up with incoming orders.
- Our kitchen (CPU) isn't working too hard — we're using our resources efficiently.

Downscaling is all about reducing resources without affecting the quality of service. It's about ensuring that we're not overstaffed when the rush hour is over.

#### **Summary: A Recipe for Effective Scaling**

In summary, our scaling policy is for scale up, we first ensure that the time to drain the backlog is beyond the threshold (120s) or the cpu is above the threshold (90%)

Increasing Backlog aka Backlog Growth > 0 :

> $`Worker_require = Worker_current \frac{Backlog Growth + Throughput}{Throughput}`$

Reviewer: Ah, assuming this is what you were talking about - I'm guessing this just isn't supported; you could take a look through other blogs or website areas and see how they handle it. I'm guessing they just triple backtick

Consistent Backlog aka Backlog Growth = 0:

$`Worker_extra = Worker_current \frac{Backlog Time}{Time to Reduce Backlog}`$

To Sum up:

$`Worker_scaleup = min(Worker_require + Worker_extra, Worker_max)`$

To scale down, we need to ensure the machine utilization is low (< 70%) and there is no backlog growth and current time to drain backlog is less than the limit (10s)

So the only driving factor to calculate the required resources after a scale down is CPU

$`CPURate_desired = \frac{Worker_current}{Worker_new} CpuRate_current`$

## Executing Autoscaling Decision

> In Our Setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As Adviced most of Flink documentation We set per vCPU one slot. Most of Our jobs uses 1 vCPU 4GB Memory combination for TaskManager.

Reviewer: ```suggestion
In our setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As Advised in most of the Flink documentation we set one per vCPU slot. Most of our jobs use 1 vCPU 4GB Memory combination for TaskManager.
```

<img class="center-block"
src="/images/blog/apache-beam-flink-and-kubernetes-part3/adaptive_scheduler_rescale.png"
alt="How Reactive Mode Works">

Parallelism can only be influenced by setting a maximum for pipelines, which the scheduler will honor. Our maxParallelism is limited by the total count of partitions that the pipeline will process, as well as by the job itself. We cap the maximum number of TaskManagers with maxWorker count and control the job's key count in shuffle by setting maxParallelism. Additionally, we set maxParallelism per pipeline to manage pipeline parallelism. The job cannot exceed the job's maxParallelism in terms of workers.

After autoscaler analysis, we will tag if the job needs to be scaled up, no action or scaled down. To interact with the job, we use a library we have built over Flink Kubernetes Operator. This library allows us to interact with our flink jobs via a simple java method call. Library will convert our method call to a kubernetes command.

In the kubernetes world, the call will look like this for a scale up:

`kubectl scale flinkdeployment job-name --replicas=100`

> Rest of thing will be handled by Apache Flink.

Reviewer: ```suggestion
Apache Flink will handle the rest of the work needed to scale up.
```

<img class="center-block"
src="/images/blog/apache-beam-flink-and-kubernetes-part3/flink-operator-chaining.png"
alt="Apache Beam Pipeline Optimization by Apache Flink">

##### **Upscaling Signals**

##### *Backlog Growth*
Let’s take a practical example. Consider a pipeline reading from Kafka, where different operators handle data parsing, formatting, and accumulation. The key metric here is throughput — how much data each operstor processes over time. But throughput alone isn't enough. We need to examine the queue size or backlog at each operator. A growing backlog indicates we're falling behind. We measure this as backlog growth — the first derivative of backlog size over time, highlighting our processing deficit.

```math

> Backlog Growth = \frac{d(Backlog)}{dt}

Reviewer: Seems like this is close to what we want, but it is still not quite right - https://apache-beam-website-pull-requests.storage.googleapis.com/29860/blog/apache-beam-flink-and-kubernetes-part3/index.html

Author: Yes. According docsy's document it should work but it does not https://www.docsy.dev/docs/adding-content/diagrams-and-formulae/

Yesterday I tried to build on my local. But I am getting this error. I could not investigate more.

```> Task :website:initGitSubmodules FAILED
fatal: detected dubious ownership in repository at '/opt'
To add an exception for this directory, call:

        git config --global --add safe.directory /opt
```
        
 Do you have any idea how can I build the website ?

Reviewer: Have you tried to run the suggested command? `git config --global --add safe.directory /opt`

Unless you're operating on a shared drive that should be safe - https://stackoverflow.com/questions/72978485/git-submodule-update-failed-with-fatal-detected-dubious-ownership-in-repositor - it is a git issue, not an issue with the website building itself

---
title:  "Behind the Scenes: Crafting an Autoscaler for Apache Beam in a High-Volume Streaming Environment"
date:   2023-12-21 09:00:00 -0400

> math: true

Reviewer: Looks like this did not work. Could you try writing the equations in ways that don't require this special math block?

Increasing Backlog aka Backlog Growth > 0 :

```math
Worker_{require} = Worker_{current} \frac{Backlog Growth + Throughput}{Throughput}
```

Consistent Backlog aka Backlog Growth = 0:

```math
Worker_{extra} = Worker_{current} \frac{Backlog Time}{Time to Reduce Backlog}
```

To Sum up:

```math
Worker_{scaleup} = min(Worker_{require} + Worker_{extra}, Worker_{max})
```

To scale down, we need to ensure the machine utilization is low (< 70%) and there is no backlog growth and current time to drain backlog is less than the limit (10s)

So the only driving factor to calculate the required resources after a scale down is CPU

```math
CPURate_{desired} = \frac{Worker_{current}}{Worker_{new}} CpuRate_{current}
```

## Executing Autoscaling Decision

> In our setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As Advised in most of the Flink documentation we set one per vCPU slot. Most of our jobs use 1 vCPU 4GB Memory combination for TaskManager.

Reviewer: ```suggestion
In our setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As advised in most of the Flink documentation we set one per vCPU slot. Most of our jobs use 1 vCPU 4GB Memory combination for TaskManager.
```

Apache Flink will handle the rest of the work needed to scale up.

## Maintaining State for Stateful Streaming Application with Autoscaling

Adapting Apache Flink's state recovery mechanisms for autoscaling involves leveraging its robust features like max parallelism, checkpointing, and the Adaptive Scheduler to ensure efficient and resilient stream processing, even as the system dynamically adjusts to varying loads. Here's how these components work together in an autoscaling context:

1. **Max Parallelism** sets an upper limit on how much a job can scale out, ensuring that state can be redistributed across a larger or smaller number of nodes without exceeding predefined boundaries. This is crucial for autoscaling because it allows Flink to manage state effectively, even as the number of task slots changes to accommodate varying workloads.

> 2. **Checkpointing** is at the heart of Flink's fault tolerance mechanism, periodically saving the state of each job to a durable storage(In our case it is GCS bucket). In an autoscaling scenario, checkpointing enables Flink to recover to a consistent state after scaling operations. When the system scales out (adds more resources) or scales in (removes resources), Flink can restore the state from these checkpoints, ensuring data integrity and processing continuity without losing critical information. In scale down or up situations there could be a moment to reprocess data from last checkpoint. To reduce that amount we reduce checkpointing interval to 10 seconds.

Reviewer: ```suggestion
2. **Checkpointing** is at the heart of Flink's fault tolerance mechanism, periodically saving the state of each job to a durable storage (in our case it is GCS bucket). In an autoscaling scenario, checkpointing enables Flink to recover to a consistent state after scaling operations. When the system scales out (adds more resources) or scales in (removes resources), Flink can restore the state from these checkpoints, ensuring data integrity and processing continuity without losing critical information. In scale down or up situations there could be a moment to reprocess data from last checkpoint. To reduce that amount we reduce the checkpointing interval to 10 seconds.
```
